src/drivers/uart/src/amdmmio.rs

```diff
@@ -117,7 +117,7 @@ impl Driver for AMDMMIO {
     fn pwrite(&mut self, data: &[u8], _offset: usize) -> Result<usize> {
         'outer: for (sent_count, &c) in data.iter().enumerate() {
             for _ in 0..RETRY_COUNT {
-                if !self.stat.is_set(STAT::THRE) || !self.stat.is_set(STAT::THRE) {
+                if self.stat.is_set(STAT::THRE) || self.stat.is_set(STAT::THRE) {
```

Comment:
This looks a bit odd to me - as it did before... Is it intended that `THRE` (transmitter holding register empty) is being checked twice, and the formerly defined `TDRE`, which is now `TEMT` (transmitter empty), is not actually used here? Or should the second check actually look at `TEMT`?

Having not too much of an understanding, I am reading the following in a datasheet:

> THRE-flag = "1" indicates that there is no data in the Transmission FIFO buffer register and transmission character is able to be written.
> TEMT-flag becomes "1" when there is no data in the register and Transmission shift register in the
transmission control part.
> Both flags become "0" at writing "0" to transmission FIFO buffer.

Reply:
>This looks a bit odd to me - as it did before... Is it intended that THRE (transmitter holding register empty) is being checked twice

Ron did that part, and I kept it in there. Maybe there's a weird quirk in the AMD hardware (I haven't seen this quirk before)? Or it's in order to make the loop do more and thus keep the CPU pipeline more full? The latter seems useless since I doubt that the UART is pipelined, too, or as fast as the CPU (so it would matter) in the first place ;)

Anyway, I'll ask him.

>, and the formerly defined TDRE, which is now TEMT (transmitter empty), is not actually used here? 

>Or should the second check actually look at TEMT?

No.

There's a parallel to serial conversion going on in the 16550.

I can only speculate on how, but the standard way to do this is to have an array of D flip flops like that (that is NOT THR--rather it's the shift register which the THR goes into): https://www.globalspec.com/ImageRepository/LearnMore/20166/digital%20parallel%20serial%20convert%20diagram%20fig%2078d6e819311dc46368055b18a318367ac.png

Q0...Q3 are the parallel input bits (in our case those Q_i are connected to THR's outputs). (We currently use 8 bits, though--and actually, it's configurable how many bits in the parallel serial converter we want in the 16550).

The transmission process goes like this:

First step, all the MUXes are set to their bottom input.

Within one rising clock edge [A], all the flop flips will load their respective Q_i into their respective internal storage.

Then, all the MUXes are set to their top input.

For all the other clock edges, for each flip flop, it loads the value from the flip flop to the left of it, respectively, on the rising clock edge. Because of the synchronous design each flip flop will get the data of the previous flip flop with a delay of 1 cycle. That means you only have to wait at the rightmost flip flop's output in order to get the output to put on the serial line.

In any case, standard serial is probably a tiny little more involved than that (adds start and stop bits)--but the idea is the same.

`THRE` signals that the first clock edge [A] is done and the data has been loaded into the parallel serial converter. It does not mean that the parallel serial converter is done actually putting them on the serial line.

`TEMT`signals that the parallel serial converter is done actually putting all the bits on the serial line and is now "empty".

As the data producer, we don't care whether the parallel serial converter is done, we just care whether buffer (THR) space is available or not. if we did check `TEMT`, then that means we'd always wait until the parallel to serial converter is done, which would cause an additional delay of at least 1 cycle (more if the stop bit is done internally by the parallel to serial converter).

So I would not check `TEMT`--it's useless information to us--and also really a hardware implementation detail.

Even if we wanted to make sure that the signal made it to the receiver, then it's also useless, since the receiver probably has a 16550, too, and their receiving serial to parallel converter has a delay, too.

In short, I don't see why `TEMT` is there in the hardware at all. Maybe for debugging it...

Reply:
You should remove it, if the contents on the left of get evaluated to true, the ones on the right don't get evaluated. If you need to check twice to make sure, then you should loop. The code wasn't right before.

Reply:
It's not to make sure, it's to keep the pipeline full in the face of a branch misprediction--if it has any use, that is.

Reply:
I've removed it anyway--if we need it, we can re-add it.